spec/integration/application/agent_spec.rb

```diff
@@ -122,6 +122,43 @@
       end
     end
 
+    it "re-evaluates a deferred function in a cached catalog" do
+      Puppet[:report] = false
+      Puppet[:use_cached_catalog] = true
+      Puppet[:usecacheonfailure] = false
+
+      catalog_dir = File.join(Puppet[:client_datadir], 'catalog')
+      Puppet::FileSystem.mkpath(catalog_dir)
+      cached_catalog_path = "#{File.join(catalog_dir, Puppet[:certname])}.json"
+
+      # our catalog contains a deferred function that calls `binary_file`
+      # to read `source`. The function returns a Binary object, whose
+      # base64 value is printed to stdout
+      source = tmpfile('deferred_source')
+      catalog = File.read(my_fixture('cached_deferred_catalog.json'))
+      catalog.gsub!('__SOURCE_PATH__', source)
+      File.write(cached_catalog_path, catalog)
```

Comment:
Alternatively, it would be easier to use `TimeStamp` and make sure it's different on the second run (assuming TimeStamp has sufficient resolution, looking at you Ruby on Windows)

```puppet
notify { "deferred":
  message => Deferred("new", [String, Deferred("new", [TimeStamp])])
}
```